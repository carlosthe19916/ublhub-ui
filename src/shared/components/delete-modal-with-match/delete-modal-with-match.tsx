import React from "react";
import { useState, useEffect } from "react";
import {
  Button,
  Modal,
  ButtonVariant,
  ModalVariant,
  TextInput,
} from "@patternfly/react-core";

export interface DeleteModalWithMatchProps {
  isModalOpen: boolean;

  title: string;
  message: string;

  matchText: string;

  inProgress?: boolean;
  onDelete: () => void;
  onCancel?: () => void;
}

export const DeleteModalWithMatch: React.FC<DeleteModalWithMatchProps> = ({
  isModalOpen,
  title,
  message,
  matchText,
  inProgress,
  onDelete,
  onCancel,
}) => {
  const [inputMatchValue, setInputMatchValue] = useState("");
  const [allowDeletion, setAllowDeletion] = useState(false);

  useEffect(() => {
    setAllowDeletion(matchText.toLowerCase() === inputMatchValue.toLowerCase());
  }, [inputMatchValue, matchText]);

  const handleDelete = () => {
    onDelete();
  };

  const handleCancel = () => {
    if (onCancel) {
      onCancel();
    }
  };

  return (
    <Modal
      variant={ModalVariant.small}
      title={title}
      isOpen={isModalOpen}
      onClose={handleCancel}
      aria-label="delete-with-match-modal"
      actions={[
        <Button
          key="delete"
          variant={ButtonVariant.danger}
          isDisabled={!allowDeletion || inProgress}
          onClick={handleDelete}
        >
          Eliminar
        </Button>,
        <Button
          key="cancel"
          variant="link"
          onClick={handleCancel}
          isDisabled={inProgress}
        >
          Cancelar
        </Button>,
      ]}
    >
      <div className="pf-c-content">
        <p>{message}</p>

        <p>Escribe '{matchText}' para confirmar.</p>

        <TextInput
          type="text"
          id="matchText"
          name="matchText"
          value={inputMatchValue}
          onChange={(value) => setInputMatchValue(value)}
          aria-label="Text input match"
          autoComplete="off"
        />
      </div>
    </Modal>
  );
};
