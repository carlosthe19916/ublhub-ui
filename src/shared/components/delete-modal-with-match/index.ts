export { DeleteModalWithMatch } from "./delete-modal-with-match";
