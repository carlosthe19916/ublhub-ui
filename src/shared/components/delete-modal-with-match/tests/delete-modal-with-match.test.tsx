import React from "react";
import { mount, shallow } from "enzyme";
import { DeleteModalWithMatch } from "../delete-modal-with-match";
import { Button } from "@patternfly/react-core";

describe("DeleteModalWithMatch", () => {
  it("Renders without crashing", () => {
    const wrapper = shallow(
      <DeleteModalWithMatch
        isModalOpen={true}
        title="Delete modal title"
        message="Are you sure you want to delete this?"
        matchText="delete"
        onDelete={jest.fn}
        onCancel={jest.fn}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Valid cancel", () => {
    const mockOnCancel = jest.fn();
    const wrapper = mount(
      <DeleteModalWithMatch
        isModalOpen={true}
        title="Delete modal title"
        message="Are you sure you want to delete this?"
        matchText="delete"
        onDelete={jest.fn}
        onCancel={mockOnCancel}
      />
    );

    const cancelButton = wrapper.find(Button).at(2);
    expect(cancelButton.props().children).toEqual("Cancelar");

    wrapper.find(Button).at(2).simulate("click");
    expect(mockOnCancel).toHaveBeenCalledTimes(1);
  });

  it("Activate 'delete' button and then callback to delete", () => {
    const matchText = "match this text";

    const mockOnDelete = jest.fn();

    const wrapper = mount(
      <DeleteModalWithMatch
        isModalOpen={true}
        title="Delete modal title"
        message="Are you sure you want to delete this?"
        matchText={matchText}
        inProgress={false}
        onDelete={mockOnDelete}
        onCancel={jest.fn()}
      />
    );

    // Verify action buttons
    const deleteButton = wrapper.find(Button).at(1);
    const cancelButton = wrapper.find(Button).at(2);

    expect(deleteButton.props().children).toEqual("Eliminar");
    expect(cancelButton.props().children).toEqual("Cancelar");

    expect(deleteButton.props().isDisabled).toEqual(true);
    expect(cancelButton.props().isDisabled).toEqual(false);

    // Match text
    const nameInput = wrapper.find('input[name="matchText"]');
    (nameInput.getDOMNode() as HTMLInputElement).value = matchText;
    nameInput.simulate("change");
    wrapper.update();

    expect(wrapper.find(Button).at(1).props().isDisabled).toEqual(false);

    // Click delete
    wrapper.find(Button).at(1).simulate("click");
    expect(mockOnDelete).toHaveBeenCalledTimes(1);
  });
});
