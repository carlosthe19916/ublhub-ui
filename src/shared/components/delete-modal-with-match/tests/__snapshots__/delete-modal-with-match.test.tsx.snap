// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteModalWithMatch Renders without crashing 1`] = `
<Modal
  actions={
    Array [
      <Button
        isDisabled={true}
        onClick={[Function]}
        variant="danger"
      >
        Eliminar
      </Button>,
      <Button
        onClick={[Function]}
        variant="link"
      >
        Cancelar
      </Button>,
    ]
  }
  appendTo={[Function]}
  aria-describedby=""
  aria-label="delete-with-match-modal"
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={false}
  isOpen={true}
  onClose={[Function]}
  ouiaSafe={true}
  showClose={true}
  title="Delete modal title"
  titleIconVariant={null}
  titleLabel=""
  variant="small"
>
  <div
    className="pf-c-content"
  >
    <p>
      Are you sure you want to delete this?
    </p>
    <p>
      Escribe '
      delete
      ' para confirmar.
    </p>
    <TextInput
      aria-label="Text input match"
      autoComplete="off"
      id="matchText"
      name="matchText"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
</Modal>
`;
